### PR TITLE
@acjay: Increment Policy

### PR DESCRIPTION
**Pull request by ashfurrow**
This is a WIP because it still needs tests. We will likely have to amend the metaphysics sample JSON fixture we use.

Re: https://github.com/artsy/eigen/issues/1719
See also: https://github.com/artsy/gravity/pull/10133 & https://github.com/artsy/metaphysics/pull/320

### CODE REVIEW

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionBidViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    case (.BidBecameMaxBidder, .BidBecameMaxBidder): return true
+    case (.BidAcknowledged, .BidAcknowledged): return true
+    case (.BidNetworkFail, .BidNetworkFail): return true
+    case (.LotWaitingToOpen, .LotWaitingToOpen): return true
+    case (.LotSold, .LotSold): return true
+    case (.BidOutbid, .BidOutbid): return true
+
+    default: return false
+    }
+}
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionBidViewModel.swift]: Everything above this line was copied directly from `LiveAuctionPlaceMaxBidViewController.swift`, and the `LiveAuctionBidViewModel` class is similar but modified. Line 95 onwards is new.

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift [hunk 2]
```diff
@@ -68,6 +71,12 @@ class LiveAuctionsSalesPerson: NSObject, LiveAuctionsSalesPersonType {
         self.stateManager = stateManagerCreator(host: host, sale: sale, saleArtworks: self.lots, jwt: jwt, bidderCredentials: biddingCredentials)
         self.auctionViewModel = auctionViewModelCreator(sale: sale, currentLotSignal: stateManager.currentLotSignal, biddingCredentials: biddingCredentials)
     }
+
+    lazy var bidIncrements: [BidIncrementStrategy] = {
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift]: A salesperson is like a view model for the sale, but is specifically concerned with interactive aspects, like what the current lot is and how to bid. It seemed like storing the (sorted) increments here made the most sense.

File: Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift [hunk 3]
```diff
@@ -128,6 +127,10 @@ class LiveAuctionLotViewModel: NSObject, LiveAuctionLotViewModelType {
         return model.artwork.dimensionsCM
     }
 
+    var askingPrice: UInt64 {
+        return model.askingPriceCents
+    }
+
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/ViewModels/LiveAuctionLotViewModel.swift]: Would have preferred to keep this exposed only as an `Observable`, but we've run into several places now where we need the asking price at-hand.

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">No CHANGELOG changes made</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Needs testing on a Phone if change is non-trivial</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Detected some Swift building time outliers</td>
    </tr>
  </tbody>
</table>

Time | Class | Function |
| --- | ----- | ----- |245.1ms | [AuctionViewController.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Artsy/View_Controllers/Auction/AuctionViewController.swift#L183) | @objc dynamic func showRefineTappedAnimated(animated: Bool)
194.4ms | [SwiftyJSON.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Pods/SwiftyJSON/Source/SwiftyJSON.swift#L445) | get {}
173.2ms | [WebSocket.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Pods/Starscream/Source/WebSocket.swift#L528) | private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
79.2ms | [LiveAuctionsAdminViewController.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Artsy/View_Controllers/Live_Auctions/LiveAuctionsAdminViewController.swift#L22) | @objc override func viewDidLoad()
76.5ms | [Observable.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Pods/Interstellar/Sources/Observable.swift#L71) | private final func nextTokenHash() -> Int
73.1ms | [AuctionLotMetadataStackScrollView.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Artsy/View_Controllers/Live_Auctions/Views/AuctionLotMetadataStackScrollView.swift#L18) | required init(viewModel: LiveAuctionLotViewModelType, salesPerson: LiveAuctionsSalesPersonType, sideMargin: String)
64.2ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L25) | func ar_presentBlurredOverlayWithTitle(title: String, subtitle: String, buttonState: BlurredStatusOverlayViewCloseButtonState = default)
62.8ms | [LiveAuctionLotViewController.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift#L49) | @objc override func viewDidLoad()
62.2ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L31) | (closure)
57.9ms | [RefinementOptionsViewController+Private.swift](/artsy/eigen/tree/ashfurrow-bid-increments/Artsy/View_Controllers/Auction/RefinementOptionsViewController+Private.swift#L63) | func stackView() -> ORStackView


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift [hunk 4]
```diff
@@ -134,6 +143,14 @@ extension LiveAuctionsSalesPerson {
     func leaveMaxBidOnLot(lot: LiveAuctionLotViewModelType, amountCents: UInt64, biddingViewModel: LiveAuctionBiddingViewModelType) {
         stateManager.leaveMaxBidOnLot(lot.lotID, amountCents: amountCents, biddingViewModel: biddingViewModel)
     }
+
+    func currentLotValue(lot: LiveAuctionLotViewModelType) -> UInt64 {
```
Comment (acjay) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift]: I'm a bit confused what this represents, is this an increment above the asking price?

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionBidViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    var canMakeLowerBids: Bool {
+        return currentBid - increments.minimumNextBidCentsIncrement(currentBid)  >= currentLotValue
+    }
+
+    func nextBidCents(bid: UInt64) -> UInt64 {
+        return bid + increments.minimumNextBidCentsIncrement(bid)
+    }
+
+    func previousBidCents(bid: UInt64) -> UInt64 {
+        return bid - increments.minimumNextBidCentsIncrement(bid)
```
Comment (acjay) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionBidViewModel.swift]: I might be misunderstanding, but it seems like `nextBidCents` and `previousBidCents` might not quite be opposites of one another. It seems like if they straddle a threshold, toggling back and forth might get you off-increment. I wonder if it makes sense just to iterate all the increments from the asking price and store that locally? It's a cheap calculation and just a couple hundred bytes of memory.

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift (hunk 4 above)
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionSalesPerson.swift]: Yeah, it's name is a bit confusing. I'd like to rename it when I get a chance.

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionBidViewModel.swift (hunk 5 above)
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionBidViewModel.swift]: Hmm, yeah it's worth more thought than I can give before I've had coffee. Enumerating them upfront would be worth it for the https://github.com/artsy/eigen/issues/1579 , the max bid UI changes. What would be a good upper bound? And they'd have to be recomputed when the asking price changes, too.

Comment (acjay) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionBidViewModel.swift]: Don't know what Force is using...but I think the original idea was 3 * max(currentPrice, highEstimate). But it doesn't matter too much.

It actually wouldn't be totally insane to just enumerate up from zero, just to have something that basically works. We should never end up off-increment in the pre-bidding.

**Comment by ashfurrow** [comment]
Okay, I've pushed a commit that precomputes the bid increments and makes things easier for the upcoming UI change (noted in the code). I also fixed an existing, undiagnosed bug where the "decrease increment" button was enabled even on the lowest increment (this is why the snapshots changed).

**Comment by acjay** [comment]
I'm no Eigen expert, but it looks good to me. Feel free to proceed!

**Comment by ashfurrow** [comment]
Cool, gonna /cc @orta on this so he can take a look when he gets a chance.

**Pull request merged**